python/pyiceberg/catalog/rest.py

```diff
@@ -0,0 +1,435 @@
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {409: TableAlreadyExistsError})
+
+        table_response = TableResponse(**response.json())
+
+        return Table(
+            identifier=(self.name,) + self.identifier_to_tuple(identifier),
+            metadata_location=table_response.metadata_location,
+            metadata=table_response.metadata,
+        )
+
+    def list_tables(self, namespace: Optional[Union[str, Identifier]] = None) -> List[Identifier]:
+        namespace_concat = NAMESPACE_SEPARATOR.join(self.identifier_to_tuple(namespace or ""))
+        response = requests.get(
+            self.url(Endpoints.list_tables, namespace=namespace_concat),
+            headers=self.headers,
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {404: NoSuchNamespaceError})
+        response = ListTablesResponse(**response.json())
+
+        return [(self.name, *entry.namespace, entry.name) for entry in response.identifiers]
+
+    def load_table(self, identifier: Union[str, Identifier]) -> Table:
+        response = requests.get(
+            self.url(Endpoints.load_table, prefixed=True, **self._split_identifier_for_path(identifier)), headers=self.headers
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {404: NoSuchNamespaceError})
+
+        table_response = TableResponse(**response.json())
+
+        return Table(
+            identifier=(self.name,) + self.identifier_to_tuple(identifier),
+            metadata_location=table_response.metadata_location,
+            metadata=table_response.metadata,
+        )
+
+    def drop_table(self, identifier: Union[str, Identifier], purge_requested: bool = False) -> None:
+        response = requests.delete(
+            self.url(Endpoints.drop_table, prefixed=True, purge=purge_requested, **self._split_identifier_for_path(identifier)),
+            headers=self.headers,
+        )
+        try:
+            response.raise_for_status()
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {404: NoSuchNamespaceError})
+
+    def purge_table(self, identifier: Union[str, Identifier]) -> None:
+        self.drop_table(identifier=identifier, purge_requested=True)
+
+    def rename_table(self, from_identifier: Union[str, Identifier], to_identifier: Union[str, Identifier]):
+        payload = {
+            "source": self._split_identifier_for_json(from_identifier),
+            "destination": self._split_identifier_for_json(to_identifier),
+        }
+        response = requests.post(self.url(Endpoints.rename_table), json=payload, headers=self.headers)
+        try:
+            response.raise_for_status()
+        except HTTPError as exc:
+            self._handle_non_200_response(exc, {404: NoSuchNamespaceError, 409: TableAlreadyExistsError})
```

Comment:
This should also be `NoSuchTableError`: https://github.com/apache/iceberg/blob/master/open-api/rest-catalog-open-api.yaml#L752-L753

Looks like the spec has an incorrect example there, but the 404 for this route means there was no such table to rename.

Reply:
I actually came up with this by just creating a faulty call in the Java impl

Reply:
Yeah, 404 is context-specific. My comment was for the `list_tables` method, where a 404 indicates that the namespace does not exist and can't be listed. That shouldn't raise `NoSuchTableError` because 404 doesn't mean the namespace was empty -- that would result in an empty list of identifiers -- 404 means that the namespace itself doesn't exist.

Sorry for the confusion, I know it's a bit odd to have the meaning of 404 change based on the route, but it makes a bit more sense when you think about routes as identifying resources. `/namespace/ns/tables` primarily identifies a namespace and is interacting with the table set of that namespace.